tpl_lib/dxf/dxf.tpl

```diff
@@ -149,56 +149,148 @@ function bounds_center(bounds) {
 }
 
 
-function quad_bezier(p, t) {
-  var c = [sqr(1 - t), 2 * (1 - t) * t, sqr(t)];
+// Find the spline segment for the t value
+function segment_get_index(spl, knots, t) {
+  // TODO Can this be unified?
+  if (spl.isClosed) {
+    for (let segment = 0; segment < knots.length; segment++)
+      if (t < knots[segment]) return segment - 1
 
-  return {
-    x: c[0] * p[0].x + c[1] * p[1].x + c[2] * p[2].x,
-    y: c[0] * p[0].y + c[1] * p[1].y + c[2] * p[2].y
+  } else
+    for (let segment = spl.degree; segment < knots.length - 1; segment++)
+      if (knots[segment] <= t && t <= knots[segment + 1])
+        return segment
+
+  throw new Error('segment not found');
+}
+
+
+function nurbs_build_verts(spl) {
+  var degree = spl.degree
+  var npts   = spl.ctrlPts.length
+
+  if (degree < 1) throw new Error('degree must be at least 1 (linear)')
+  if (npts - 1 < degree)
+    throw new Error('degree must be less than or equal to point count - 1')
+
+  // Convert control points to homogeneous coordinates
+  var verts = []
+  let nverts = spl.isClosed ? npts + degree * 2 : npts
+
+  for (let i = 0; i < nverts; i++) {
+    let j = spl.isClosed ? (i - degree + npts) % npts : i
+    let weight = spl.isRational ? spl.ctrlPts[j].w : 1
+
+    verts.push([
+      spl.ctrlPts[j].x * weight,
+      spl.ctrlPts[j].y * weight,
+      spl.ctrlPts[j].z * weight,
+      weight
+    ])
   }
+
+  return verts
 }
 
 
-function quad_bezier_length(p) {
-  var l = 0;
-  var v = p[0];
+function nurbs_build_knots(spl) {
+  var degree = spl.degree
+  var npts   = spl.ctrlPts.length
+
+  if (spl.isClosed) {
+    // TODO Is it correct to rewrite the knots values?
```

Comment:
Good question.  Why cant we use the knot vector supplied in the DXF file.  I suppose some DXF files may be missing the knot vector.  This code assumes a uniform knot vector which may not agree with the DXF.

Reply:
The [bspline-curve-closed](https://pages.mtu.edu/~shene/COURSES/cs3621/NOTES/spline/B-spline/bspline-curve-closed.html) page has two methods to close the nurb. I implemented 'Wrapping Control Points', but it also has a 'Wrapping Knots' method that mentions 'These knots are not necessarily uniform, an advantage over the method discussed above', so maybe that's more correct when the dxf specifies control points?
If you want, you can give it a try and see if works better.

Reply:
I'd like to figure this out.  The DXF files I've seen contain knot vectors in this format:

    0,0,0,d,2*d,3*d,. . .,(n-1)*d,1,1,1

Where ``d`` is the uniform knot delta.  Obviously in a closed, periodic spline you don't need the leading 0's and trailing 1's but how to you convert the knot vector?  Part of the problem is the number of knots to control points in a DXF file does not quite make sense to me.  I suppose we could just forego non-uniform splines and always generate a uniform knot vector.  Perhaps non-uniform splines don't often occur in DXF files.  We could use some more DXF file spline examples.  Really we need a test suite of DXF inputs.

Reply:
Same here... I diffed a few open and closed splines from LibreCAD and all the knots/control_points were the same; the only diff was in the isClosed flag.
Maybe we can try to take a look at LibreCAD's rendering code since it's open source?
But yeah, I'm not sure how often non-uniform splines are encountered in CADs.
An option could be to assert that the knots are in that uniform format you described above and pop-up a warning when someone tries to load non-uniform splines, but asserting on floating point values is always a bit risky...

Reply:
I did already look at LibreCAD's spline code.  It uses a library called opennurbs.  In the process of calling in to opennurbs, LibreCAD also appears to throw out the knot vector supplied by the DXF file.